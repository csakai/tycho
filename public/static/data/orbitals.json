[{
	"name": "Earth",
	"albedo": 0.367,
	"sidereal": 0.99726968,
	"radius": 6371,
	"markdown": "earth.md",
	"mass": 5.97237e+24,
	"centralMass": 1.989e+30,
	"axialTilt": 23.4392811,
	"inclination": 7.155,
	"atmosphere": 35260,
	"satellites": [{
		"name": "Moon",
		"albedo": 0.136,
		"sidereal": 0.99726968,
		"radius": 1737.1,
		"markdown": "moon.md",
		"mass": 5.97237e+24,
		"centralMass": 5.97237e+24,
		"axialTilt": 6.687,
		"inclination": 5.145,
		"atmosphere": 10066278,
		"ephemeris": {
			"nasaOrbitalCode": "301",
			"nasaBarycenterCode": "3",
			"revolutionOrder": "d"
		},
		"argumentOfPeriapsis": 282.5992409518417,
		"eccentricity": 0.03405532141489297,
		"longitudeOfAscendingNode": 142.3595722910625,
		"semimajor": 376126.5103019063,
		"semiminor": 375908.33786092995,
		"periapses": {
			"last": 1509927058416.626,
			"next": 1512377208118.1555
		},
		"id": "moon"
	}],
	"ephemeris": {
		"nasaOrbitalCode": "3",
		"nasaBarycenterCode": "0",
		"revolutionOrder": "y"
	},
	"argumentOfPeriapsis": 114.20763,
	"eccentricity": 0.01355360061112275,
	"longitudeOfAscendingNode": 348.73936,
	"semimajor": 150759239.9570005,
	"semiminor": 150745392.07806867,
	"periapses": {
		"last": 1136430000000,
		"next": 1167987600000
	},
	"id": "earth"
},{
	"name": "Venus",
	"albedo": 0.689,
	"sidereal": -243.025,
	"radius": 6051.8,
	"markdown": "venus.md",
	"mass": 4.8675e+24,
	"centralMass": 1.989e+30,
	"axialTilt": 2.64,
	"inclination": 3.86,
	"atmosphere": 16052641,
	"ephemeris": {
		"nasaOrbitalCode": "2",
		"nasaBarycenterCode": "0",
		"revolutionOrder": "y"
	},
	"argumentOfPeriapsis": 55.186,
	"eccentricity": 0.0067,
	"longitudeOfAscendingNode": 76.678,
	"semimajor": 108208179,
	"semiminor": 107477094,
	"periapses": {
		"last": 1429348560000,
		"next": 1448787300000
	},
	"id": "venus"
}]