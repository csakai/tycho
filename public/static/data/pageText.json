{"planetScale":"Planet Scale: × ","speedScale":"Speed Scale: × 10^","settings":"Settings"}
