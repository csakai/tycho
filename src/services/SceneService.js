import TWEEN from 'tween.js';
import {Vector3} from 'three';

export default class SceneService {

  /**
   * Converts an instance of Vector3 to a plain object with x, y, z properties.
   *
   * @param {Vector3} v - vector to convert
   * @returns {Object} - plain object with x, y, z properties
   */
  static vectorToObject = (v) => {
    const {x, y, z} = v;
    return {x, y, z};
  }

  /**
   * Converts a plain object to an instance of Vector3.
   *
   * @param {Object} o - object, with x, y, z properties
   * @returns {Vector3} - converted vector
   */
  static objectToVector = (o) => {
    return new Vector3(o.x, o.y, o.z);
  }

  /**
   * Runs a given tween instance with the provided params.
   *
   * @param {Tween} tween - instance of tween
   * @param {Vector3} dest - destination vector
   * @param {Function} onComplete - callback
   */
  static startTween = (tween, dest, onComplete) => {
    return tween
      .easing(TWEEN.Easing.Quadratic.Out)
      .to(SceneService.vectorToObject(dest), 2000)
      .onComplete(onComplete)
      .start();
  }

  /**
   * Returns the vector of the active target position.
   * If no target position is found, it will return <0>.
   *
   * @param {Object} state - scene state
   * @param {Object} props - scene props
   * @returns {Vector3} - active target vector
   */
  static getTargetPosition = (state, props) => {
    if (state) {
      const {positions} = state;
      const {perspective, targetName} = props;

      if (positions[targetName] && !perspective) {
        return positions[targetName].position3d;
      }
    }
    return new Vector3(0, 0, 0);
  }

  /**
   * Updates the camera position and lookAt vectors for the given target positions.
   *
   * @param {Object} state - scene state
   * @param {Object} props - scene props
   * @param {Camera} camera - scene camera
   */
  static updateCameraVectors = (state, props, camera) => {
    const {positions, targetName} = state;
    const {perspective, lookAtName} = props;

    if(positions[targetName] && positions[lookAtName] && perspective) {
      camera.position.copy(positions[targetName].position3d);
      camera.lookAt(positions[lookAtName].position3d);
    }
  }

  /**
   * Gets the zoom delta from controls and invokes the given action if zoom has changed.
   *
   * @param {Event} event - DOM mousewheel event
   * @param {Controls} controls - instance of controls
   * @param {Function} action - callback action
   */
  static mapZoom = (event, controls, action) => {
    const zoom = controls.getZoomDelta(event.deltaY);
    const current = Math.round(controls.level * 100);

    if (current !== zoom) {
      action(zoom);
    }
  }
}
