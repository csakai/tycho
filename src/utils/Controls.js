import * as THREE from 'three';
import OrbitControls from 'three-orbit-controls';
import TWEEN from 'tween.js';
import Constants from '../constants';

export default class Controls extends OrbitControls(THREE) {

  constructor(camera, domElement) {
    super(camera, domElement);

    this.camera = camera;
    this.enabled = true;
    this.enableZoom = false;
    this.enablePan = false;
    this.level = Constants.WebGL.Zoom.MAX;
    this.minDistance = Constants.WebGL.Camera.MIN_DISTANCE;
    this.maxDistance = Constants.WebGL.Camera.MAX_DISTANCE;
  }

  /**
   * Change zoom. Overrides the existing zoom tween in progress.
   *
   * @param {level} level - percentage of desired zoom level [0,100]
   */
  zoom = (level) => {
    if(this.level !== level) {
      this.pan(level / 100);
      this.level = level;
    }
  }

  /**
   * Gets the zoom delta from controls and invokes the given action if zoom has changed.
   *
   * @param {Event} ev - DOM wheel event
   * @param {Function} action - callback action
   */
  wheelZoom = (ev, action) => {
    const zoom = this.getZoomDelta(ev.deltaY);
    const current = Math.round(this.level * 100);

    if (current !== zoom) {
      action(zoom);
    }
  }

  /**
   * Calculates the zoom percentage from the given mousewheel delta.
   *
   * @param {Number} delta - mousewheel delta
   * @returns {Number} new zoom level
   */
  getZoomDelta = (delta) => {
    let zoom = this.level;
    
    zoom += (delta / this.getWheelDeltaDivisor(zoom));
    zoom = Math.max(zoom, 0);
    zoom = Math.min(zoom, 100);
   
    return zoom;
  }

  /**
   * Calculates a wheel delta that inflates based on how far away
   * the user is from the target, because outer space is really big.
   *
   * @param {Number} level - current zoom level 
   * @returns {Number} scrollwheel delta
   */
  getWheelDeltaDivisor = (level) => {
    let delta = Constants.UI.WHEEL_DELTA_DIVISOR;
    let match = false;

    Constants.WebGL.ScrollScale
      .forEach(({distance, scale}) => {
        if (level > distance && !match) {
          delta *= scale;
          match = true;
        }
      });

    return delta;
  }

  /**
   * Sets the current camera position to the scaled zoom vector.
   *
   * @param {Number} percent - percentage of zoom [0,1]
   */
  pan = (percent, log) => {
    let position = this.camera.position;
    let newVector = this.getZoomVector(position, this.maxDistance * percent);
    let minVector = this.getZoomVector(position, this.minDistance);
    
    if (newVector.length() >= this.minDistance) {
      position.copy(newVector);
    } else {
      position.copy(minVector);
    }
  }

  /**
   * Returns the current vector scaled to the desired magnitude.
   *
   * @param {Vector3} vector - current position vector
   * @param {Number} scalar - magnitude of vector
   * @returns {Vector3}
   */
  getZoomVector = (vector, scalar) => {
    return vector
      .clone()
      .normalize()
      .multiplyScalar(scalar);
  }

  /**
   * Returns the distance between <0> and current camera position.
   * 
   * @returns {Number} distance
   */
  getDistance = () => {
    return this.maxDistance - this.minDistance;
  }

  /**
   * Enable controls.
   */
  enable = () => {
    this.enabled = true;
  }

  /**
   * Disable controls.
   */
  disable = () => {
    this.enabled = false;
  }

  /**
   * Starts autorotating the scene clockwise.
   *
   * @param {Number} speed - speed to rotate at
   */
  startAutoRotate = (speed) => {
    this.autoRotate = true;
    this.autoRotateSpeed = speed;
  }

  /**
   * Stop autorotating the scene.
   */
  stopAutoRotate = () => {
    this.autoRotate = false;
  }

  /**
   * Deletes Tween instance and data.
   */
  endTween = () => {
    delete this.tweenData;
    delete this.tweenBase;
  }

  /**
   * Cancels Tween, if one is in progress.
   */
  cancelTween = () => {
    if (this.tweenBase) {
      this.tweenBase.stop();
      this.endTween();
    }
  }

  /**
   * Zooms to the level of the current Tween data.
   */
  updateTween = () => {
    this.zoom(this.tweenData.level);
  }

  /**
   * Invokes teardown methods for active tween.
   */
  completeTween = () => {
    if (this.tweenDone) {
      this.tweenDone(this.level);
    }
    this.endTween();
  }

  /**
   * Creates an animation to zoom to the specified level.
   *
   * @param {Number} level - percentage of zoom
   * @param {Function} onDone - animation completion callback
   */
  tweenZoom = (level, onDone) => {
    this.cancelTween();

    this.tweenDone = onDone;
    this.tweenData = {level: this.level};

    this.tweenBase = new TWEEN
      .Tween(this.tweenData)
      .easing(TWEEN.Easing.Quadratic.Out)
      .to({level}, Constants.WebGL.Tween.SLOW)
      .onUpdate(this.updateTween)
      .onComplete(this.completeTween)
      .start();
  }
}
