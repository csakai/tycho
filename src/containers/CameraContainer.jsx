import React from 'react';
import TWEEN from 'tween.js';
import SceneService from '../services/SceneService';
import {Vector3} from 'three';

const cameraPosition = new Vector3(300, 300, 300);//move to const

export default class CameraContainer extends React.Component {

  componentDidMount = () => {
    if (this.props.cameraRef) {
      this.props.cameraRef(this.refs.camera);
    }
  }

  componentWillReceiveProps = (nextProps) => {
    if (this.props.targetName !== nextProps.targetName) {
      this.startTween(nextProps.targetName);
    }
  }

  /**
   * Removes any active Tween.
   */
  endTween = () => {
    delete this.tweenData;
    delete this.tweenBase;
  }

  /**
   * Cancels any Tween in progress.
   */
  cancelTween = () => {
    if (this.tweenBase) {
      this.tweenBase.stop();
      this.endTween();
    }
  }

  /**
   * Starts Tween to the position having a key of targetName.
   */
  startTween = (targetName) => {
    const target = this.props.positions[targetName];

    if (target) {
      this.cancelTween();
      this.assignTween();
      this.props.zoomIn();

      SceneService.startTween(this.tweenBase, target.position3d, this.endTween);
    }
  }

  /**
   * Assigns a new Tween to the active target position.
   */
  assignTween = () => {
    this.tweenData = SceneService.vectorToObject(this.getTargetPosition());
    this.tweenBase = new TWEEN.Tween(this.tweenData);
  }
  
  /**
   * Returns the calculated active target position.
   * If a Tween is in progress, it will return that current position.
   */
  getTargetPosition = () => {
    const {positions, targetName} = this.props;

    if (this.tweenData) {
      return SceneService.objectToVector(this.tweenData);
    }
    return SceneService.getTargetPosition(positions, targetName);
  }

  render() {
    return (
      <group position={this.getTargetPosition()}>
        <perspectiveCamera
          name="camera"
          ref="camera"
          fov={50}
          aspect={window.innerWidth / window.innerHeight}
          near={1}
          far={10000}
          position={cameraPosition}
        />
      </group>
    );
  }
}

