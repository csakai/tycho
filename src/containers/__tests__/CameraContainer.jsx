import React from 'react';
import React3 from 'react-three-renderer';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import TWEEN from 'tween.js';
import {Vector3, Camera} from 'three';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import CameraContainer from '../CameraContainer';
import Controls from '../../utils/Controls';
import data from '../../global/fixtures';
import SceneService from '../../services/SceneService';

jest.mock('../../services/SceneService');

describe('Camera Container', () => {
  let component, cameraContainer;

  beforeEach(() => {
    component = shallow(
      <CameraContainer 
        cameraRef={jest.fn()}
        positions={{}}
        targetName="testPlanet"
        zoomIn={jest.fn()}
        controls={{}}
      />
    );
    cameraContainer = component.instance();
  });

  describe('componentDidMount()', () => {
    it('should pass the camera reference to the cameraRef prop callback', () => {
      cameraContainer.props = {cameraRef: jest.fn()};
      const spy = jest.spyOn(cameraContainer.props, 'cameraRef');

      cameraContainer.refs = {camera: new Camera()};
      cameraContainer.componentDidMount();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(cameraContainer.refs.camera);
    });

    it('should not call the cameraRef prop callback if undefined', () => {
      const spy = jest.spyOn(cameraContainer.props, 'cameraRef');

      cameraContainer.props = {};
      cameraContainer.componentDidMount();

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('componentWillReceiveProps()', () => {
    it('should start tweening to `targetName` if the prop has changed', () => {
      const controls = new Controls(new Camera());
      const targetName = 'Mars';
      const nextProps = {targetName: 'Earth'};
      const positions = {};
      const spy = jest.spyOn(cameraContainer, 'startTween');

      cameraContainer.props = {targetName, controls, positions};
      cameraContainer.componentWillReceiveProps(nextProps);

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(nextProps.targetName);
    });

    it('should not tween if the `targetName` if the prop has not changed', () => {
      const controls = new Controls(new Camera());
      const targetName = 'Mars';
      const nextProps = {targetName};
      const positions = {};
      const spy = jest.spyOn(cameraContainer, 'startTween');

      cameraContainer.props = {targetName, controls, positions};
      cameraContainer.componentWillReceiveProps(nextProps);

      expect(spy).not.toHaveBeenCalled();
    });
  });
 
  describe('endTween()', () => {
    it('should dispose of the tweenData and tweenBase', () => {
      cameraContainer.tweenData = {};
      cameraContainer.tweenBase = {};
      cameraContainer.endTween();

      expect(cameraContainer).not.toHaveProperty('tweenData');
      expect(cameraContainer).not.toHaveProperty('tweenBase');
    });
  });

  describe('cancelTween()', () => {
    it('should call endTween, if the tweenBase is defined', () => {
      const stop = jest.fn();
      const spy = jest.spyOn(cameraContainer, 'endTween');

      cameraContainer.tweenBase = {stop};
      cameraContainer.cancelTween();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });

  describe('startTween()', () => {
    const targetName = 'Earth';
    const action = {};
    const cameraBase = {};
    const zoomIn = jest.fn();
    const positions = {[targetName]: {}};

    beforeEach(() => {
      cameraContainer.refs = {cameraBase};
      cameraContainer.props = {action, positions, zoomIn};
    });

    it('should call cancelTween()', () => {
      const spy = jest.spyOn(cameraContainer, 'cancelTween');

      cameraContainer.startTween(targetName);
      
      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should initialize a new tween', () => {
      cameraContainer.startTween(targetName);
      expect(typeof cameraContainer.tweenBase).toEqual('object');
    });
  });

  describe('getTargetPosition()', () => {
    it('should return the vector version of current tweenData, if any', () => {
      const tweenData = {x: 1, y: 1, z: 1};

      SceneService.objectToVector = (v) => v;
      cameraContainer.tweenData = tweenData;

      const result = cameraContainer.getTargetPosition();

      expect(result).toEqual(tweenData);
    });

    it('should return the target position if scene container is not actively tweening', () => {
      const targetPosition = {x: 1, y: 1, z: 1};
      SceneService.getTargetPosition = () => targetPosition;
      const result = cameraContainer.getTargetPosition();

      expect(result).toEqual(targetPosition);
    });
  });

  describe('render()', () => {
    it('should render the scene container successfully', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
