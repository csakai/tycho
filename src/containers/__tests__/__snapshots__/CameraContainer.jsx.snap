// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Camera Container render() should render the scene container successfully 1`] = `
<group
  position={
    Object {
      "x": 1,
      "y": 1,
      "z": 1,
    }
  }
>
  <perspectiveCamera
    aspect={1.3333333333333333}
    far={10000}
    fov={50}
    name="camera"
    near={1}
    position={
      Vector3 {
        "x": 300,
        "y": 300,
        "z": 300,
      }
    }
  />
</group>
`;
